pkg/trace/sampler/remote_rates.go

```diff
@@ -0,0 +1,225 @@
+package sampler
+
+import (
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config/remote/service"
+	"github.com/DataDog/datadog-agent/pkg/proto/pbgo"
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/watchdog"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// RemoteRates sharded by signature. Allowing an independent feedback
+// loop per signature. RemoteRates adjusts sampling rates communicated
+// to the agent based on the observed traffic. It targets a TPS configured
+// remotely.
+type RemoteRates struct {
+	samplers   map[Signature]*Sampler
+	tpsTargets map[Signature]float64
+	mu         sync.RWMutex // protects concurrent access to samplers and tpsTargets
+
+	exit    chan struct{}
+	stopped chan struct{}
+}
+
+func newRemoteRates(conf *config.AgentConfig) *RemoteRates {
+	if !conf.RemoteRates {
+		return nil
+	}
+
+	remoteRates := &RemoteRates{
+		samplers: make(map[Signature]*Sampler),
+
+		exit:    make(chan struct{}),
+		stopped: make(chan struct{}),
+	}
+
+	err := service.NewGRPCSubscriber(
+		pbgo.Product_APM_SAMPLING,
+		func(config *pbgo.ConfigResponse) error {
+			log.Infof("fetched config version %d from remote config management", config.DirectoryTargets.Version)
+			return remoteRates.loadNewConfig(config)
+		},
+	)
+	if err != nil {
+		log.Errorf("Error when subscribing to remote config management %v", err)
+		return nil
+	}
+	return remoteRates
+}
+
+func (r *RemoteRates) loadNewConfig(config *pbgo.ConfigResponse) error {
+	tpsTargets := make(map[Signature]float64, len(r.tpsTargets))
+	for _, targetFile := range config.TargetFiles {
+		var rates pb.RemoteRates
+		_, err := rates.UnmarshalMsg(targetFile.Raw)
+		if err != nil {
+			return err
+		}
+		for _, rate := range rates.Rates {
+			sig := ServiceSignature{Name: rate.Service, Env: rate.Env}.Hash()
+			tpsTargets[sig] = rate.Rate
+		}
+	}
+	r.updateTPS(tpsTargets)
+	return nil
+}
+
+func (r *RemoteRates) updateTPS(tpsTargets map[Signature]float64) {
+	r.mu.Lock()
+	r.tpsTargets = tpsTargets
+	r.mu.Unlock()
+
+	// update samplers with new TPS
+	r.mu.RLock()
+	noTPSConfigured := map[Signature]struct{}{}
+	for sig, sampler := range r.samplers {
+		rate, ok := tpsTargets[sig]
+		if !ok {
+			noTPSConfigured[sig] = struct{}{}
+		}
+		sampler.UpdateTargetTPS(rate)
+	}
+	r.mu.RUnlock()
+
+	// trim signatures with no TPS configured
+	r.mu.Lock()
+	for sig := range noTPSConfigured {
+		delete(r.samplers, sig)
+	}
+	r.mu.Unlock()
+}
+
+// Start runs and adjust rates per signature following remote TPS targets
+func (r *RemoteRates) Start() {
+	go func() {
+		defer watchdog.LogOnPanic()
```

Comment:
Not sure, but should we make sure that this section of code cannot be started multiple times?